File: sdn-controller/lib/topology/network_graph.ml
(* Network Graph Implementation
   =============================
   This module models the trading firm's network as a weighted directed graph.

   KEY DESIGN DECISIONS (for interview):

   1. We use OCamlGraph under the hood — it's a mature graph library with
      efficient Dijkstra implementation. We don't reinvent graph algorithms.

   2. The graph is IMMUTABLE (functional updates return new graphs). This is
      critical for concurrent access: the telemetry collector can update metrics
      while the path computation engine reads the graph, without locks or races.
      In a mutable graph, you'd need locking, which adds latency to path
      computation — unacceptable for a trading network controller.

   3. We use a PERSISTENT (immutable) graph rather than imperative because:
      - Thread safety without locks (see above)
      - Easy snapshotting: we can keep the "before" and "after" graph when
        a link fails, useful for debugging and the chaos framework
      - Functional style matches OCaml idioms — this is Jane Street's language

   4. The weight function for Dijkstra is parameterized — the caller decides
      what "shortest" means. This lets us optimize for different objectives:
      - Minimum latency (for trading traffic)
      - Maximum bandwidth (for bulk data transfers)
      - Composite cost (latency + utilization penalty) for the SDN controller
*)

(* ============================================================
   Types
   ============================================================ *)

type device_role =
  | CoreRouter
  | SpineSwitch
  | LeafSwitch
  | EdgeRouter
  | TransitRouter
  | ExchangeRouter
[@@deriving show, eq]

type device = {
  id : string;
  name : string;
  role : device_role;
  site : string;
  loopback : string;
  vendor : string;
  asn : int;
}
[@@deriving show, eq]

type link_state =
  | Up
  | Down
  | Degraded
[@@deriving show, eq]

type link_metrics = {
  latency_us : float;
  jitter_us : float;
  utilization : float;
  packet_loss : float;
  last_updated : float;
}
[@@deriving show, eq]

type link = {
  id : string;
  src_interface : string;
  dst_interface : string;
  bandwidth_mbps : int;
  state : link_state;
  metrics : link_metrics;
  ospf_cost : int;
}
[@@deriving show, eq]

(* ============================================================
   Internal Graph Representation

   We use two maps:
   - devices: id -> device
   - adjacency: src_id -> (dst_id -> link) map

   This is a simple adjacency-list representation using immutable maps.
   We COULD use OCamlGraph's built-in persistent graph, but rolling our
   own adjacency map gives us more control over the API and makes the
   code more transparent for interview discussion.

   For path computation, we DO use OCamlGraph's Dijkstra — no reason
   to reimplement a well-tested algorithm.
   ============================================================ *)

module StringMap = Map.Make(String)

type t = {
  devices : device StringMap.t;
  adjacency : link StringMap.t StringMap.t;
  (* adjacency.(src).(dst) = link from src to dst *)
}

(* ============================================================
   Graph Construction
   ============================================================ *)

let empty () = {
  devices = StringMap.empty;
  adjacency = StringMap.empty;
}

let add_device (graph : t) (device : device) : t =
  if StringMap.mem device.id graph.devices then
    invalid_arg (Printf.sprintf "Device %s already exists" device.id)
  else
    { devices = StringMap.add device.id device graph.devices;
      adjacency = StringMap.add device.id (StringMap.empty : link StringMap.t) graph.adjacency;
    }

let add_link graph ~src ~dst link =
  if not (StringMap.mem src graph.devices) then
    raise (Not_found)
  else if not (StringMap.mem dst graph.devices) then
    raise (Not_found)
  else
    let src_adj =
      StringMap.find src graph.adjacency
      |> StringMap.add dst link
    in
    (* Add the link in both directions — network links are bidirectional.
       This is important: if DC-East connects to DC-West, traffic flows both ways. *)
    let dst_adj =
      StringMap.find dst graph.adjacency
      |> StringMap.add src link
    in
    { graph with
      adjacency =
        graph.adjacency
        |> StringMap.add src src_adj
        |> StringMap.add dst dst_adj;
    }

(* ============================================================
   Queries
   ============================================================ *)

let find_device graph id =
  StringMap.find_opt id graph.devices

let find_link graph ~src ~dst =
  match StringMap.find_opt src graph.adjacency with
  | None -> None
  | Some adj -> StringMap.find_opt dst adj

let all_devices graph =
  StringMap.bindings graph.devices |> List.map snd

let all_links graph =
  StringMap.fold (fun src adj acc ->
    StringMap.fold (fun dst link acc ->
      (* Only include each link once (src < dst) to avoid duplicates
         since we store links bidirectionally *)
      if String.compare src dst < 0 then
        (src, dst, link) :: acc
      else
        acc
    ) adj acc
  ) graph.adjacency []

let neighbors graph id =
  match StringMap.find_opt id graph.adjacency with
  | None -> []
  | Some adj ->
    StringMap.bindings adj
    |> List.filter (fun (_dst, link) ->
      (* Only return neighbors reachable via Up or Degraded links.
         Down links are effectively non-existent for routing. *)
      match link.state with
      | Up | Degraded -> true
      | Down -> false)

let devices_at_site graph site =
  all_devices graph
  |> List.filter (fun d -> String.equal d.site site)

(* ============================================================
   Metric Updates

   These return NEW graphs (immutable update). The old graph is
   still valid — this is how we get thread safety without locks.
   ============================================================ *)

let update_link_in_adj adj dst f =
  match StringMap.find_opt dst adj with
  | None -> adj
  | Some link -> StringMap.add dst (f link) adj

let update_metrics (graph : t) ~src ~dst (metrics : link_metrics) : t =
  let update_adj a b adjacency =
    match StringMap.find_opt a adjacency with
    | None -> adjacency
    | Some adj ->
      StringMap.add a (update_link_in_adj adj b (fun link ->
        { link with metrics }
      )) adjacency
  in
  { graph with
    adjacency =
      graph.adjacency
      |> update_adj src dst
      |> update_adj dst src;
  }

let update_link_state (graph : t) ~src ~dst (new_state : link_state) : t =
  let update_adj a b adjacency =
    match StringMap.find_opt a adjacency with
    | None -> adjacency
    | Some adj ->
      StringMap.add a (update_link_in_adj adj b (fun link ->
        { link with state = new_state }
      )) adjacency
  in
  { graph with
    adjacency =
      graph.adjacency
      |> update_adj src dst
      |> update_adj dst src;
  }

(* ============================================================
   Path Computation — Dijkstra's Algorithm

   This is the core of the SDN controller's intelligence.

   WHY DIJKSTRA (for interview):
   - Dijkstra computes the shortest path in O((V + E) log V) time
   - With our topology size (~30 devices, ~60 links), this runs in
     microseconds — well within our 50ms decision budget
   - The weight function is parameterized, so we can optimize for
     different objectives without changing the algorithm
   - We use a priority queue (min-heap) implementation

   WHY NOT Bellman-Ford?
   - Bellman-Ford handles negative weights, but network metrics
     (latency, cost) are never negative
   - Bellman-Ford is O(V * E), slower than Dijkstra for our case
   - We'd use Bellman-Ford if we needed to detect negative cycles
     (not applicable in networking)
   ============================================================ *)

(* Simple priority queue using a sorted list.
   For our topology size (~30 nodes), this is fast enough.
   For larger topologies, we'd use a proper binary heap. *)
module PQ = struct
  type 'a t = (float * 'a) list

  let empty : 'a t = []

  let push (pq : 'a t) priority item : 'a t =
    let rec insert = function
      | [] -> [(priority, item)]
      | ((p, _) as hd) :: tl ->
        if priority <= p then (priority, item) :: hd :: tl
        else hd :: insert tl
    in
    insert pq

  let pop = function
    | [] -> None
    | (priority, item) :: rest -> Some (priority, item, rest)
end

let shortest_path graph ~src ~dst ~weight =
  (* Dijkstra's algorithm:
     1. Start at src with distance 0
     2. For each unvisited node, pick the one with smallest distance
     3. Update distances to its neighbors through current node
     4. Repeat until we reach dst or exhaust all reachable nodes *)
  let rec dijkstra pq visited prev dist =
    match PQ.pop pq with
    | None -> None  (* No path exists — network partition *)
    | Some (current_dist, current, pq') ->
      if String.equal current dst then begin
        (* Found the destination — reconstruct the path *)
        let rec build_path node acc =
          match StringMap.find_opt node prev with
          | None -> acc  (* Reached the source *)
          | Some (prev_node, link) ->
            build_path prev_node ((node, link) :: acc)
        in
        Some (build_path dst [], current_dist)
      end else if StringMap.mem current visited then
        (* Already visited this node — skip *)
        dijkstra pq' visited prev dist
      else begin
        let visited = StringMap.add current () visited in
        (* Explore all neighbors *)
        let pq', prev', dist' =
          neighbors graph current
          |> List.fold_left (fun (pq, prev, dist) (neighbor, link) ->
            let edge_weight = weight link in
            let new_dist = current_dist +. edge_weight in
            let old_dist =
              match StringMap.find_opt neighbor dist with
              | Some d -> d
              | None -> Float.infinity
            in
            if new_dist < old_dist then
              ( PQ.push pq new_dist neighbor,
                StringMap.add neighbor (current, link) prev,
                StringMap.add neighbor new_dist dist )
            else
              (pq, prev, dist)
          ) (pq', prev, dist)
        in
        dijkstra pq' visited prev' dist'
      end
  in
  let pq = PQ.push PQ.empty 0.0 src in
  let dist = StringMap.singleton src 0.0 in
  dijkstra pq StringMap.empty StringMap.empty dist

let all_shortest_paths graph ~src ~weight =
  (* Run Dijkstra from src, but collect ALL destinations *)
  let rec dijkstra pq visited prev dist results =
    match PQ.pop pq with
    | None -> results  (* Exhausted all reachable nodes *)
    | Some (current_dist, current, pq') ->
      if StringMap.mem current visited then
        dijkstra pq' visited prev dist results
      else begin
        let visited = StringMap.add current () visited in
        (* Reconstruct path to current node *)
        let rec build_path node acc =
          match StringMap.find_opt node prev with
          | None -> acc
          | Some (prev_node, link) ->
            build_path prev_node ((node, link) :: acc)
        in
        let results =
          if not (String.equal current src) then
            (current, build_path current [], current_dist) :: results
          else
            results
        in
        let pq', prev', dist' =
          neighbors graph current
          |> List.fold_left (fun (pq, prev, dist) (neighbor, link) ->
            let edge_weight = weight link in
            let new_dist = current_dist +. edge_weight in
            let old_dist =
              match StringMap.find_opt neighbor dist with
              | Some d -> d
              | None -> Float.infinity
            in
            if new_dist < old_dist then
              ( PQ.push pq new_dist neighbor,
                StringMap.add neighbor (current, link) prev,
                StringMap.add neighbor new_dist dist )
            else
              (pq, prev, dist)
          ) (pq', prev, dist)
        in
        dijkstra pq' visited prev' dist' results
      end
  in
  let pq = PQ.push PQ.empty 0.0 src in
  let dist = StringMap.singleton src 0.0 in
  dijkstra pq StringMap.empty StringMap.empty dist []

(* ============================================================
   JSON Serialization

   The REST API and dashboard consume this to display the
   network state in real time.
   ============================================================ *)

let device_role_to_string = function
  | CoreRouter -> "core_router"
  | SpineSwitch -> "spine_switch"
  | LeafSwitch -> "leaf_switch"
  | EdgeRouter -> "edge_router"
  | TransitRouter -> "transit_router"
  | ExchangeRouter -> "exchange_router"

let device_role_of_string = function
  | "core_router" -> Ok CoreRouter
  | "spine_switch" -> Ok SpineSwitch
  | "leaf_switch" -> Ok LeafSwitch
  | "edge_router" -> Ok EdgeRouter
  | "transit_router" -> Ok TransitRouter
  | "exchange_router" -> Ok ExchangeRouter
  | s -> Error (Printf.sprintf "Unknown device role: %s" s)

let link_state_to_string = function
  | Up -> "up"
  | Down -> "down"
  | Degraded -> "degraded"

let link_state_of_string = function
  | "up" -> Ok Up
  | "down" -> Ok Down
  | "degraded" -> Ok Degraded
  | s -> Error (Printf.sprintf "Unknown link state: %s" s)

let device_to_yojson (d : device) =
  `Assoc [
    ("id", `String d.id);
    ("name", `String d.name);
    ("role", `String (device_role_to_string d.role));
    ("site", `String d.site);
    ("loopback", `String d.loopback);
    ("vendor", `String d.vendor);
    ("asn", `Int d.asn);
  ]

let metrics_to_yojson (m : link_metrics) =
  `Assoc [
    ("latency_us", `Float m.latency_us);
    ("jitter_us", `Float m.jitter_us);
    ("utilization", `Float m.utilization);
    ("packet_loss", `Float m.packet_loss);
    ("last_updated", `Float m.last_updated);
  ]

let link_to_yojson (l : link) =
  `Assoc [
    ("id", `String l.id);
    ("src_interface", `String l.src_interface);
    ("dst_interface", `String l.dst_interface);
    ("bandwidth_mbps", `Int l.bandwidth_mbps);
    ("state", `String (link_state_to_string l.state));
    ("metrics", metrics_to_yojson l.metrics);
    ("ospf_cost", `Int l.ospf_cost);
  ]

let to_yojson graph =
  let devices =
    all_devices graph
    |> List.map device_to_yojson
  in
  let links =
    all_links graph
    |> List.map (fun (src, dst, link) ->
      `Assoc [
        ("src", `String src);
        ("dst", `String dst);
        ("link", link_to_yojson link);
      ])
  in
  `Assoc [
    ("devices", `List devices);
    ("links", `List links);
  ]

(* JSON deserialization helpers *)
let json_string = function
  | `String s -> Ok s
  | _ -> Error "Expected string"

let json_int = function
  | `Int i -> Ok i
  | _ -> Error "Expected int"

let json_float = function
  | `Float f -> Ok f
  | `Int i -> Ok (Float.of_int i)
  | _ -> Error "Expected float"

let json_field assoc key =
  match List.assoc_opt key assoc with
  | Some v -> Ok v
  | None -> Error (Printf.sprintf "Missing field: %s" key)

let ( let* ) = Result.bind

(* Helper: extract field then apply a conversion *)
let field_as conv assoc key =
  let* v = json_field assoc key in
  conv v

let device_of_yojson = function
  | `Assoc assoc ->
    let* id = field_as json_string assoc "id" in
    let* name = field_as json_string assoc "name" in
    let* role_str = field_as json_string assoc "role" in
    let* role = device_role_of_string role_str in
    let* site = field_as json_string assoc "site" in
    let* loopback = field_as json_string assoc "loopback" in
    let* vendor = field_as json_string assoc "vendor" in
    let* asn = field_as json_int assoc "asn" in
    Ok { id; name; role; site; loopback; vendor; asn }
  | _ -> Error "Expected JSON object for device"

let metrics_of_yojson = function
  | `Assoc assoc ->
    let* latency_us = field_as json_float assoc "latency_us" in
    let* jitter_us = field_as json_float assoc "jitter_us" in
    let* utilization = field_as json_float assoc "utilization" in
    let* packet_loss = field_as json_float assoc "packet_loss" in
    let* last_updated = field_as json_float assoc "last_updated" in
    Ok { latency_us; jitter_us; utilization; packet_loss; last_updated }
  | _ -> Error "Expected JSON object for metrics"

let link_of_yojson = function
  | `Assoc assoc ->
    let* id = field_as json_string assoc "id" in
    let* src_interface = field_as json_string assoc "src_interface" in
    let* dst_interface = field_as json_string assoc "dst_interface" in
    let* bandwidth_mbps = field_as json_int assoc "bandwidth_mbps" in
    let* state_str = field_as json_string assoc "state" in
    let* state = link_state_of_string state_str in
    let* metrics_json = json_field assoc "metrics" in
    let* metrics = metrics_of_yojson metrics_json in
    let* ospf_cost = field_as json_int assoc "ospf_cost" in
    Ok { id; src_interface; dst_interface; bandwidth_mbps; state; metrics; ospf_cost }
  | _ -> Error "Expected JSON object for link"

let of_yojson = function
  | `Assoc assoc ->
    let* devices_json = json_field assoc "devices" in
    let* links_json = json_field assoc "links" in
    let* device_list = match devices_json with
      | `List l -> List.fold_left (fun acc j ->
          let* acc = acc in
          let* d = device_of_yojson j in
          Ok (d :: acc)
        ) (Ok []) l
      | _ -> Error "Expected array for devices"
    in
    let graph = List.fold_left (fun g d -> add_device g d) (empty ()) device_list in
    let* _link_list = match links_json with
      | `List l -> List.fold_left (fun acc j ->
          let* acc = acc in
          match j with
          | `Assoc link_assoc ->
            let* src = field_as json_string link_assoc "src" in
            let* dst = field_as json_string link_assoc "dst" in
            let* link_json = json_field link_assoc "link" in
            let* link = link_of_yojson link_json in
            Ok ((src, dst, link) :: acc)
          | _ -> Error "Expected JSON object for link entry"
        ) (Ok []) l
      | _ -> Error "Expected array for links"
    in
    let graph = List.fold_left (fun g (src, dst, link) ->
      add_link g ~src ~dst link
    ) graph _link_list in
    Ok graph
  | _ -> Error "Expected JSON object for graph"

File: sdn-controller/tradenet_sdn/network_graph.py
"""Network Graph — models the trading firm's network topology as a weighted graph.

This is the core data structure of the SDN controller. Every routing decision
starts here: the topology graph represents all devices (nodes) and links (edges)
with their current metrics (latency, utilization, packet loss).

Why a graph? Networks ARE graphs — routers are nodes, links are edges. Modeling
them explicitly lets us run graph algorithms (Dijkstra) to compute optimal paths,
rather than relying on distributed routing protocols alone.

Design notes (ported from the original OCaml implementation):

  1. The graph is IMMUTABLE. Every "update" returns a NEW graph and leaves the
     original untouched. This is critical for concurrent access: the telemetry
     collector can update metrics while the path-computation engine reads the
     graph, without locks or races. In Python the GIL makes the ref-swap atomic;
     the immutable snapshot means a reader always sees a consistent graph.

  2. We keep two maps:
       devices:   id -> Device
       adjacency: src_id -> {dst_id -> Link}
     a simple adjacency-list representation. Links are stored in BOTH directions
     because network links are bidirectional.

  3. The weight function for Dijkstra is parameterized — the caller decides what
     "shortest" means (minimum latency, maximum bandwidth, composite cost, ...).
"""

from __future__ import annotations

import heapq
from dataclasses import dataclass, replace
from enum import Enum
from typing import Callable, Optional

# ============================================================
# Types
# ============================================================


class DeviceRole(str, Enum):
    """The role a device plays in the topology. The string value is the JSON form."""

    CORE_ROUTER = "core_router"      # Cisco IOSv — backbone/WAN interconnects
    SPINE_SWITCH = "spine_switch"    # Arista vEOS — data center spine
    LEAF_SWITCH = "leaf_switch"      # Arista vEOS — data center leaf
    EDGE_ROUTER = "edge_router"      # Juniper vSRX — colo perimeter with firewall
    TRANSIT_ROUTER = "transit_router"   # External transit provider (simulated)
    EXCHANGE_ROUTER = "exchange_router"  # Exchange-side router (NYSE/CME/NASDAQ)


class LinkState(str, Enum):
    """Current operational state of a link. The string value is the JSON form."""

    UP = "up"              # Link is operational
    DOWN = "down"          # Link is down (failure or admin shutdown)
    DEGRADED = "degraded"  # Link is up but experiencing issues


@dataclass(frozen=True)
class Device:
    """A network device (node in the graph)."""

    id: str           # Unique identifier, e.g. "dc-east-core-rtr-01"
    name: str         # Human-readable name
    role: DeviceRole  # What role this device plays
    site: str         # Which site it belongs to
    loopback: str     # Loopback IP address (used for BGP peering)
    vendor: str       # "cisco", "arista", or "juniper"
    asn: int          # BGP AS number


@dataclass(frozen=True)
class LinkMetrics:
    """Real-time metrics for a link, updated by the telemetry collector."""

    latency_us: float    # One-way latency in microseconds
    jitter_us: float     # Jitter (latency variance) in microseconds
    utilization: float   # Link utilization as a fraction [0.0, 1.0]
    packet_loss: float   # Packet loss rate as a fraction [0.0, 1.0]
    last_updated: float  # Unix timestamp of last metric update


@dataclass(frozen=True)
class Link:
    """A network link (edge in the graph)."""

    id: str               # Unique link identifier
    src_interface: str    # Source interface name, e.g. "GigabitEthernet0/1"
    dst_interface: str    # Destination interface name
    bandwidth_mbps: int   # Link bandwidth in Mbps
    state: LinkState      # Current operational state
    metrics: LinkMetrics  # Current real-time metrics
    ospf_cost: int        # OSPF cost assigned to this link


# A path is an ordered list of (device_id, link) pairs from source to
# destination (the source itself is not included). A weight function maps a
# link to a float cost.
Path = list[tuple[str, Link]]
WeightFn = Callable[[Link], float]


# ============================================================
# Graph
# ============================================================


class NetworkGraph:
    """An immutable weighted graph of devices and links.

    Construct with :meth:`empty` and use the functional-update methods
    (``add_device``, ``add_link``, ``update_metrics``, ``update_link_state``),
    each of which returns a NEW graph and leaves ``self`` unchanged.
    """

    __slots__ = ("_devices", "_adjacency")

    def __init__(
        self,
        devices: Optional[dict[str, Device]] = None,
        adjacency: Optional[dict[str, dict[str, Link]]] = None,
    ) -> None:
        # These dicts are treated as immutable; functional updates copy-on-write.
        self._devices: dict[str, Device] = devices if devices is not None else {}
        self._adjacency: dict[str, dict[str, Link]] = (
            adjacency if adjacency is not None else {}
        )

    # ----- Construction -----------------------------------------------------

    @staticmethod
    def empty() -> "NetworkGraph":
        """Create an empty network graph."""
        return NetworkGraph({}, {})

    def add_device(self, device: Device) -> "NetworkGraph":
        """Add a device. Returns the updated graph.

        Raises ``ValueError`` if a device with the same ID already exists.
        """
        if device.id in self._devices:
            raise ValueError(f"Device {device.id} already exists")
        new_devices = dict(self._devices)
        new_devices[device.id] = device
        new_adjacency = dict(self._adjacency)
        new_adjacency[device.id] = {}
        return NetworkGraph(new_devices, new_adjacency)

    def add_link(self, src: str, dst: str, link: Link) -> "NetworkGraph":
        """Add a link between two devices. Both devices must already exist.

        Raises ``KeyError`` if either device ID is unknown. The link is stored
        in both directions because network links are bidirectional.
        """
        if src not in self._devices:
            raise KeyError(src)
        if dst not in self._devices:
            raise KeyError(dst)
        new_adjacency = dict(self._adjacency)
        src_adj = dict(new_adjacency[src])
        src_adj[dst] = link
        new_adjacency[src] = src_adj
        dst_adj = dict(new_adjacency[dst])
        dst_adj[src] = link
        new_adjacency[dst] = dst_adj
        return NetworkGraph(dict(self._devices), new_adjacency)

    # ----- Queries ----------------------------------------------------------

    def find_device(self, device_id: str) -> Optional[Device]:
        """Look up a device by ID. Returns ``None`` if not found."""
        return self._devices.get(device_id)

    def find_link(self, src: str, dst: str) -> Optional[Link]:
        """Look up a link between two devices. Returns ``None`` if none exists."""
        adj = self._adjacency.get(src)
        if adj is None:
            return None
        return adj.get(dst)

    def all_devices(self) -> list[Device]:
        """All devices in the graph, ordered by device id (like the OCaml Map)."""
        return [self._devices[k] for k in sorted(self._devices)]

    def all_links(self) -> list[tuple[str, str, Link]]:
        """All links as (src, dst, link).

        Each undirected link is reported once (the pair with ``src < dst``) so we
        don't double-count the bidirectional storage. Ordering matches the
        original OCaml ``all_links`` (ascending fold, prepended → reversed).
        """
        acc: list[tuple[str, str, Link]] = []
        for src in sorted(self._adjacency):
            adj = self._adjacency[src]
            for dst in sorted(adj):
                if src < dst:
                    acc.append((src, dst, adj[dst]))
        acc.reverse()
        return acc

    def neighbors(self, device_id: str) -> list[tuple[str, Link]]:
        """Neighbors of a device, reachable via an Up or Degraded link.

        Down links are effectively non-existent for routing.
        """
        adj = self._adjacency.get(device_id)
        if adj is None:
            return []
        return [
            (dst, link)
            for dst, link in sorted(adj.items())
            if link.state in (LinkState.UP, LinkState.DEGRADED)
        ]

    def devices_at_site(self, site: str) -> list[Device]:
        """All devices at a specific site."""
        return [d for d in self.all_devices() if d.site == site]

    # ----- Metric / state updates (functional, return new graphs) -----------

    def _update_link(
        self, src: str, dst: str, transform: Callable[[Link], Link]
    ) -> "NetworkGraph":
        """Apply ``transform`` to the link in BOTH directions, returning a new graph."""
        new_adjacency = dict(self._adjacency)
        for a, b in ((src, dst), (dst, src)):
            adj = new_adjacency.get(a)
            if adj is not None and b in adj:
                inner = dict(adj)
                inner[b] = transform(inner[b])
                new_adjacency[a] = inner
        return NetworkGraph(dict(self._devices), new_adjacency)

    def update_metrics(
        self, src: str, dst: str, metrics: LinkMetrics
    ) -> "NetworkGraph":
        """Update the metrics for a specific link (used by the telemetry collector)."""
        return self._update_link(src, dst, lambda link: replace(link, metrics=metrics))

    def update_link_state(
        self, src: str, dst: str, new_state: LinkState
    ) -> "NetworkGraph":
        """Update the state of a link (e.g. when the chaos framework kills a link)."""
        return self._update_link(src, dst, lambda link: replace(link, state=new_state))

    # ----- Path computation (Dijkstra) --------------------------------------

    def shortest_path(
        self, src: str, dst: str, weight: WeightFn
    ) -> Optional[tuple[Path, float]]:
        """Shortest path between two devices using Dijkstra's algorithm.

        The ``weight`` function determines what "shortest" means. Returns
        ``None`` if no path exists (network partition); otherwise
        ``(path, total_weight)`` where ``path`` is the ordered list of
        (device_id, link) pairs from source to destination.
        """
        # prev[node] = (prev_node, link_used_to_reach_node)
        prev: dict[str, tuple[str, Link]] = {}
        dist: dict[str, float] = {src: 0.0}
        visited: set[str] = set()
        # Heap entries are (distance, node). Lazy deletion: stale entries are
        # skipped when popped if the node was already visited.
        pq: list[tuple[float, str]] = [(0.0, src)]

        while pq:
            current_dist, current = heapq.heappop(pq)
            if current == dst:
                return self._reconstruct(prev, dst), current_dist
            if current in visited:
                continue
            visited.add(current)
            for neighbor, link in self.neighbors(current):
                new_dist = current_dist + weight(link)
                if new_dist < dist.get(neighbor, float("inf")):
                    dist[neighbor] = new_dist
                    prev[neighbor] = (current, link)
                    heapq.heappush(pq, (new_dist, neighbor))
        return None  # No path exists — network partition

    def all_shortest_paths(
        self, src: str, weight: WeightFn
    ) -> list[tuple[str, Path, float]]:
        """Shortest paths from ``src`` to every reachable device (a routing table)."""
        prev: dict[str, tuple[str, Link]] = {}
        dist: dict[str, float] = {src: 0.0}
        visited: set[str] = set()
        pq: list[tuple[float, str]] = [(0.0, src)]
        results: list[tuple[str, Path, float]] = []

        while pq:
            current_dist, current = heapq.heappop(pq)
            if current in visited:
                continue
            visited.add(current)
            if current != src:
                results.append((current, self._reconstruct(prev, current), current_dist))
            for neighbor, link in self.neighbors(current):
                new_dist = current_dist + weight(link)
                if new_dist < dist.get(neighbor, float("inf")):
                    dist[neighbor] = new_dist
                    prev[neighbor] = (current, link)
                    heapq.heappush(pq, (new_dist, neighbor))
        return results

    @staticmethod
    def _reconstruct(prev: dict[str, tuple[str, Link]], node: str) -> Path:
        """Walk ``prev`` back from ``node`` to the source, building the forward path."""
        path: Path = []
        while node in prev:
            prev_node, link = prev[node]
            path.append((node, link))
            node = prev_node
        path.reverse()
        return path

    # ----- Serialization ----------------------------------------------------

    def to_dict(self) -> dict:
        """Serialize the graph to a JSON-ready dict (for the REST API / dashboard)."""
        return {
            "devices": [device_to_dict(d) for d in self.all_devices()],
            "links": [
                {"src": src, "dst": dst, "link": link_to_dict(link)}
                for src, dst, link in self.all_links()
            ],
        }

    @staticmethod
    def from_dict(data: dict) -> "NetworkGraph":
        """Deserialize a graph from a dict (inverse of :meth:`to_dict`).

        Raises ``ValueError`` if the structure is malformed.
        """
        if not isinstance(data, dict):
            raise ValueError("Expected JSON object for graph")
        devices_json = data.get("devices")
        links_json = data.get("links")
        if not isinstance(devices_json, list):
            raise ValueError("Expected array for devices")
        if not isinstance(links_json, list):
            raise ValueError("Expected array for links")

        graph = NetworkGraph.empty()
        for d in devices_json:
            graph = graph.add_device(device_from_dict(d))
        for entry in links_json:
            if not isinstance(entry, dict):
                raise ValueError("Expected JSON object for link entry")
            src = entry.get("src")
            dst = entry.get("dst")
            if not isinstance(src, str) or not isinstance(dst, str):
                raise ValueError("Link entry missing src/dst")
            graph = graph.add_link(src, dst, link_from_dict(entry.get("link")))
        return graph


# ============================================================
# JSON helpers (module-level, mirroring the OCaml *_to_yojson / *_of_yojson)
# ============================================================


def device_to_dict(d: Device) -> dict:
    return {
        "id": d.id,
        "name": d.name,
        "role": d.role.value,
        "site": d.site,
        "loopback": d.loopback,
        "vendor": d.vendor,
        "asn": d.asn,
    }


def metrics_to_dict(m: LinkMetrics) -> dict:
    return {
        "latency_us": m.latency_us,
        "jitter_us": m.jitter_us,
        "utilization": m.utilization,
        "packet_loss": m.packet_loss,
        "last_updated": m.last_updated,
    }


def link_to_dict(link: Link) -> dict:
    return {
        "id": link.id,
        "src_interface": link.src_interface,
        "dst_interface": link.dst_interface,
        "bandwidth_mbps": link.bandwidth_mbps,
        "state": link.state.value,
        "metrics": metrics_to_dict(link.metrics),
        "ospf_cost": link.ospf_cost,
    }


def device_from_dict(data: dict) -> Device:
    if not isinstance(data, dict):
        raise ValueError("Expected JSON object for device")
    try:
        return Device(
            id=str(data["id"]),
            name=str(data["name"]),
            role=DeviceRole(data["role"]),
            site=str(data["site"]),
            loopback=str(data["loopback"]),
            vendor=str(data["vendor"]),
            asn=int(data["asn"]),
        )
    except KeyError as exc:
        raise ValueError(f"Missing field: {exc.args[0]}") from exc
    except ValueError as exc:
        raise ValueError(f"Invalid device: {exc}") from exc


def metrics_from_dict(data: dict) -> LinkMetrics:
    if not isinstance(data, dict):
        raise ValueError("Expected JSON object for metrics")
    try:
        return LinkMetrics(
            latency_us=float(data["latency_us"]),
            jitter_us=float(data["jitter_us"]),
            utilization=float(data["utilization"]),
            packet_loss=float(data["packet_loss"]),
            last_updated=float(data["last_updated"]),
        )
    except KeyError as exc:
        raise ValueError(f"Missing field: {exc.args[0]}") from exc


def link_from_dict(data: dict) -> Link:
    if not isinstance(data, dict):
        raise ValueError("Expected JSON object for link")
    try:
        return Link(
            id=str(data["id"]),
            src_interface=str(data["src_interface"]),
            dst_interface=str(data["dst_interface"]),
            bandwidth_mbps=int(data["bandwidth_mbps"]),
            state=LinkState(data["state"]),
            metrics=metrics_from_dict(data["metrics"]),
            ospf_cost=int(data["ospf_cost"]),
        )
    except KeyError as exc:
        raise ValueError(f"Missing field: {exc.args[0]}") from exc
